src/rnnlm/rnnlm-simple-looped.h

```diff
@@ -0,0 +1,184 @@
+// rnnlm/kaldi-rnnlm-simple-looped.h
+
+// Copyright 2017 Johns Hopkins University (author: Daniel Povey)
+//           2017 Yiming Wang
+//           2017 Hainan Xu
+
+// See ../../COPYING for clarification regarding multiple authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
+// WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+// See the Apache 2 License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef KALDI_RNNLM_SIMPLE_LOOPED_H_
+#define KALDI_RNNLM_SIMPLE_LOOPED_H_
+
+#include <vector>
+#include "base/kaldi-common.h"
+#include "gmm/am-diag-gmm.h"
+#include "hmm/transition-model.h"
+//#include "itf/decodable-itf.h"
+#include "nnet3/nnet-optimize.h"
+#include "nnet3/nnet-compute.h"
+#include "nnet3/am-nnet-simple.h"
+#include "rnnlm/rnnlm-core-compute.h"
+
+namespace kaldi {
+namespace nnet3 {
+
+// See also nnet-am-decodable-simple.h, which is a decodable object that's based
+// on breaking up the input into fixed chunks.  The decodable object defined here is based on
+// 'looped' computations, which naturally handle infinite left-context (but are
+// only ideal for systems that have only recurrence in the forward direction,
+// i.e. not BLSTMs... because there isn't a natural way to enforce extra right
+// context for each chunk.)
+
+
+// Note: the 'simple' in the name means it applies to networks for which
+// IsSimpleNnet(nnet) would return true.  'looped' means we use looped
+// computations, with a kGotoLabel statement at the end of it.
+struct RnnlmSimpleLoopedComputationOptions {
+  int32 frames_per_chunk;
+  bool debug_computation;
+  bool force_normalize;
+  NnetOptimizeOptions optimize_config;
+  NnetComputeOptions compute_config;
+  RnnlmSimpleLoopedComputationOptions():
+      frames_per_chunk(1),
+      debug_computation(false),
+      force_normalize(false) { }
+
+  void Check() const {
+    KALDI_ASSERT(frames_per_chunk > 0);
+  }
+
+  void Register(OptionsItf *opts) {
+    opts->Register("frames-per-chunk", &frames_per_chunk,
+                   "Number of frames in each chunk that is separately evaluated "
+                   "by the neural net.");
+    opts->Register("debug-computation", &debug_computation, "If true, turn on "
+                   "debug for the actual computation (very verbose!)");
+    opts->Register("force-normalize", &force_normalize, "If true, force "
```

Comment:
this explanation is not very clear IMO.  Maybe
"If true, word probabilities will be correctly normalized (otherwise the sum-to-one normalization
is approximate)"